### PR TITLE
JAMES-2319 Clearing a mail queue

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: src/site/markdown/server/manage-webadmin.md [hunk 1]
```diff
@@ -1595,13 +1595,16 @@ Response codes:
...
+##### For deleting emails in a mail queue by conditions:
 This request should have exactly one query parameter from the following list:
 
 * sender: which is a mail address (i.e. sender@james.org)
 * name: which is a string
 * recipient: which is a mail address (i.e. recipient@james.org)
 
 The mails from the given mail queue matching the query parameter will be deleted.
+##### For deleting all emails in a mail queue:
+* Pass no parameter in the url
```
Comment (chibenwa) [inline src/site/markdown/server/manage-webadmin.md]: I'm pretty sure the type of task created is different...

Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: Also I think you can improve formatting and wording here (Do we need  subsection at all?)

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 2]
```diff
@@ -875,4 +511,453 @@ public void forcingDelayedMailsDeliveryShouldActuallyChangePropertyOnMails() thr
                 assertThat(delivery.get()).isBefore(ZonedDateTime.now());
             });
     }
+
+    @Nested
+    class DeleteMail {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: Can we have Nested classes for the other REST operations as well?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: This Nested class is in my opinion still too long.

If a Nested class can have Nested class, then we should introduce the following sub-categories:

 - **DataValidation** (all 400 errors and asserts on return codes)
 - **SideEffects** (every time we assert the content of the queue)
 - **HttpBodies** (every time we assert the content of the cody)

If some other categroies makes sens then we should add them. I do believe it should help us to structure our tests... WDYT ?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 3]
```diff
@@ -875,4 +511,453 @@ public void forcingDelayedMailsDeliveryShouldActuallyChangePropertyOnMails() thr
...
+            MailAddress deletedRecipientMailAddress = new MailAddress(recipient);
+            assertThat(queue.browse())
+                .hasSize(2)
+                .allSatisfy((ManageableMailQueue.MailQueueItemView item) -> {
+                    assertThat(item.getMail().getRecipients()).doesNotContain(deletedRecipientMailAddress);
+                });
+        }
+
+        @Test
+        public void deleteMailsShouldCompleteWhenNoQueryParameters() {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: Maybe Clear related asserts could be in there own dedicated Nested class?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 4]
```diff
@@ -875,4 +511,453 @@ public void forcingDelayedMailsDeliveryShouldActuallyChangePropertyOnMails() thr
...
+
+            given()
+                .basePath(TasksRoutes.BASE)
+            .when()
+                .get(taskId + "/await")
+            .then()
+                .body("status", is("completed"));
+
+            assertThat(queue.getSize()).isEqualTo(0);
+        }
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: What happens upon deleting all mails from a mailqueue that does not exist?

**Comment by chibenwa** [comment]
```
[996890e8f9daabcac333a022b9652ff018f7822a] [ERROR] src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java:[27,8] (imports) UnusedImports: Unused import - java.util.stream.Stream.
```

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java (hunk 2 above)
Comment (trantienduchn) [reply server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: Yeah, I think so. And the sub-categories in this `DeleteMail` class, I prefer some easy to understand circumstance nested class name:
- AssertFail
- AssertComplete
- AssertBody
- AssertQueueAfterDelete

Comment (trantienduchn) [reply server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: oh but your suggestion is more applicable in other test 👍

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: I don't understand what **AssertFail** refer to. I would propose **AssertValidationFails**.

**AssertHttpBody** will be more helpful.

File: src/site/markdown/server/manage-webadmin.md [hunk 5]
```diff
@@ -1627,6 +1624,32 @@ The scheduled task will have the following type `deleteMailsFromMailQueue` and t
 }
 ```
 
+### Clearing a mail queue
```
Comment (chibenwa) [inline src/site/markdown/server/manage-webadmin.md]: You need to add the link in the navigation section :-)

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 6]
```diff
@@ -433,15 +434,18 @@ private void assertPayloadContainsDelayedEntry(Request request) {
     }
 
     private Task deleteMailsTask(ManageableMailQueue queue, Optional<MailAddress> maybeSender, Optional<String> maybeName, Optional<MailAddress> maybeRecipient) {
-        if (Booleans.countTrue(maybeSender.isPresent(), maybeName.isPresent(), maybeRecipient.isPresent()) != 1) {
+        if (countAvailableParams(maybeSender, maybeName, maybeRecipient) == 0) {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: a switch would be appropriate, don't you think ?

**Comment by chibenwa** [comment]
test this please

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**